src/gc.c

```diff
@@ -112,11 +173,34 @@ void GCContext_OnDelete(GCContext* gc) {
   }
 }
 
+void GCContext_CommonForceInvoke(GCContext* gc, RedisModuleBlockedClient* bc) {
+  if (gc->stopped) {
+    return;
+  }
+
+  GCTaskCtx *task = GCTaskCreate(gc, bc);
+  thpool_add_work(gcThreadpool_g, threadCallback, task);
+}
+
 void GCContext_ForceInvoke(GCContext* gc, RedisModuleBlockedClient* bc) {
-  BlockClients_push(&gc->bClients, bc);
-  RMUtilTimer_ForceInvoke(gc->timer);
+  GCContext_CommonForceInvoke(gc, bc);
 }
 
 void GCContext_ForceBGInvoke(GCContext* gc) {
-  RMUtilTimer_Signal(gc->timer);
+  GCContext_CommonForceInvoke(gc, DEADBEEF);
 }
+
+void GC_ThreadPoolStart() {
+  if (gcThreadpool_g == NULL) {
+    gcThreadpool_g = thpool_init(1);
+  }
+}
+
+void GC_ThreadPoolDestroy() {
+  if (gcThreadpool_g != NULL) {
+    RedisModule_ThreadSafeContextUnlock(RSDummyContext);
+    thpool_destroy(gcThreadpool_g);
```

Comment:
I think the thpool_destroy are waiting for the last task to finish but you acquire the lock so the task will not be able to finish, deadlock.

Reply:
We had a deadlock which is why we use `Unlock`